### PR TITLE
reduce memory consumption during GHMMHMM multi sequence fits

### PR DESCRIPTION
**Pull request by dizcza**
Hi, today I learned about your package, started to use it, faced the memory problem, and came up with a PR that fixes it.

I've exploited the `lengths` option and added another meaning to it. Currently, for the `GMMHMM` only. Curious users will find a way to extend my implementation to other models as well.

This also partially addresses the comment left in https://github.com/hmmlearn/hmmlearn/commit/08dee6640483cda232f7d2fcc7935d4008f4d368:

https://github.com/hmmlearn/hmmlearn/blob/0562ca65756ffb60da836eeeb1845e61767c705b/lib/hmmlearn/hmm.py#L918-L922

I got rid of the unnecessary `'centered'` arrays in the `stats` dict. If you don't want to store the `post_comp_mix` matrices in the `stats`, the logic of computing intermediate variables - `c_n` and `c_d` for the covariance - should be moved from the `_do_mstep` to `_accumulate_sufficient_statistics` function. Since this is my first PR, I've decided not to rummage through your code a lot. In either case, this should be considered in a separate PR, if you will.

Best,
Danylo

### CODE REVIEW

**Comment by anntzer** [comment]
Unless I am mistaken, chunking will also prevent the transitions at the chunking points from being taken into account?

**Comment by dizcza** [comment]
> Unless I am mistaken, chunking will also prevent the transitions at the chunking points from being taken into account?

Yes, you're correct.

**Comment by anntzer** [comment]
Then I would want some (strong) evidence that the GMMHMM implementation cannot be further improved (as suggested by the comment under GMMHMM._initialize_sufficient_statistics) before encouraging users to use an option which will give wrong results.

**Comment by dizcza** [comment]
I can improve it further (to its maximum extent) by not storing neither `post_comp_mix` nor `samples` in the stats dict as I've already pointed out. But

1. The memory consumption will still be `O(N)` where `N = size(X)` because the whole input array is required in `X`. I have an idea how to make the implementation truly chunked as some scipy/scikit-learn functions do. Since you provide the `lengths`, the users can override the `__slice__` method of the input object `X` such that `X[start: stop]` is by default a numpy view (`X` is typically a numpy array) or read from a file (user-side implementation) otherwise.
2. When the data size is really large like I had when I faced the memory bottleneck, > 10M, it's better to neglect a few transitions where the array is split than run into a memory error, isn't it?
3. Since this is my first PR, I've decided not to rummage through your code a lot. In either case, moving the logic from `_do_mstep` into `_accumulate_sufficient_statistics` should be considered in a separate PR, if you will.
4. And finally, the claim
   > before encouraging users to use an option which will give wrong results.

   is exaggerated. I've explicitly provided a test that shows that chunking yields very similar results even for medium-large arrays and therefore "wrong" is quite harsh.

**Comment by dizcza** [comment]
To achieve identical results, one should slice the array with overlapping end and stop points like that: x[0: 100], x[99: 200], x[199: 300], ... But you require the `lengths` to sum up to the shape of X. Introducing a notion of overlapping slices would require rethinking what the lengths do and perhaps substituting it with a more formal variable that would essentially lead to a breaking change.

**Comment by anntzer** [comment]
> I can improve it further (to its maximum extent) by not storing neither `post_comp_mix` nor `samples` in the stats dict as I've already pointed out. But
> 3. Since this is my first PR, I've decided not to rummage through your code a lot. In either case, moving the logic from `_do_mstep` into `_accumulate_sufficient_statistics` should be considered in a separate PR, if you will.

The improvements would be useful, and can perhaps come as a first, separate PR so that they can go in without being delayed by the rest of the discussion about docs.

> 1. The memory consumption will still be `O(N)` where `N = size(X)` because the whole input array is required in `X`. I have an idea how to make the implementation truly chunked as some scipy/scikit-learn functions do. Since you provide the `lengths`, the users can override the `__slice__` method of the input object `X` such that `X[start: stop]` is by default a numpy view (`X` is typically a numpy array) or read from a file (user-side implementation) otherwise.

If things are really O(N), I actually don't understand how this is a problem because you already require O(N) memory to specify the input, so unless you're working with datasets *just* below the memory limit of your computer, chunking won't help.  I guess, however, that what you mean is actually something like O(N x n_states x n_mix).  I don't actually have much experience there, but my guess is that Gaussian mixtures don't really scale to large n_mix anyways, even when used outside of HMMs?

> 2. When the data size is really large like I had when I faced the memory bottleneck, > 10M, it's better to neglect a few transitions where the array is split than run into a memory error, isn't it?
> 4. And finally, the claim "before encouraging users to use an option which will give wrong results." is exaggerated. I've explicitly provided a test that shows that chunking yields very similar results even for medium-large arrays and therefore "wrong" is quite harsh.

That's a difference in philosophy; I believe it is better to admit that you can't compute the right value (where "right" is, of course, only "right within the standard EM approach") rather than to guess.  However, I guess I could agree with mentioning the workaround in the docs *as long as the additional approximations it implies are clearly mentioned*.

> To achieve identical results, one should slice the array with overlapping end and stop points like that: x[0: 100], x[99: 200], x[199: 300], ... But you require the `lengths` to sum up to the shape of X. Introducing a notion of overlapping slices would require rethinking what the lengths do and perhaps substituting it with a more formal variable that would essentially lead to a breaking change.

This doesn't really get rid of the approximation anyways, because you would still additionally impose that `x[99]`, `x[199]`, etc. follow `startprob` and contribute to its estimation.

**Comment by dizcza** [comment]
> because you already require O(N) memory to specify the input

yes, that's the current approach. But what if we have data that does not fit the entire RAM at once? Then we would like to split the data into batches/chunks, process in batches, accumulate in batches, and reduce in batches. When a batch is finished, the user can provide another batch to add to the statistics being reduced in the `_do_mstep` function. By reduction, I mean summation in estimating means and the covariance matrix.
Then the memory requirements will be `O(batch_size x n_mix x ...)`. This perhaps goes beyond what the project was designed for, but I thought it'd be a nice feature to have. And yeah, I do have data > 8 GB. Perhaps, a way to go is to dive into PyTorch and/or TensorFlow, but unfortunately, their API is limited. Writing an implementation by myself... well, I lack the knowledge.

> I can improve it further (to its maximum extent) by not storing neither post_comp_mix nor samples in the stats dict

Having this done, the memory footprint will be reduced to

```
O(max(size_x * n_features, batch_size * n_features * n_mix * n_components))
```

But let's discuss this separately. Because to do this, I'll need to take the `_do_mstep`'s for-loops I've introduced in this PR and move them to the `_accumulate_sufficient_statistics` function. These are two steps, both heavy, that's why I've decided to split the optimization in two steps for the sake of easy PR review.

> This doesn't really get rid of the approximation anyways, because you would still additionally impose that x[99], x[199], etc. follow startprob and contribute to its estimation.

Hm, in my understanding, the approximation will be exact, again because of the sum in estimating the mean & cov matrix, and `_do_mstep` is called at the end of each iteration, not per batch. Therefore, keeping the reduced sums allows us to split the input data in batches and accumulate the sum at the end of each iteration.

**Comment by anntzer** [comment]
> But what if we have data that does not fit the entire RAM at once

OK, that's fair, I could not have expected that you intended to "manually" run the steps of the EM.

> Hm, in my understanding, the approximation will be exact,

Here is an admittedly artificial example:
```python
from hmmlearn.hmm import GaussianHMM
import numpy as np

# 0 100 times, 1 100 times, 0 100 times, etc.
x = (np.arange(100_00) // 100) % 2 + .1 * np.random.randn(100_00)
chunks = [x[100 * i:][:101] for i in range(100)]
chunked = np.concatenate(chunks)
hmm1 = GaussianHMM(2).fit(x[:, None])
hmm2 = GaussianHMM(2).fit(chunked[:, None], lengths=[len(chunk) for chunk in chunks])
print(hmm1.startprob_)
print(hmm1.transmat_)
print(hmm1.means_)
print(hmm1.covars_)
print(hmm2.startprob_)
print(hmm2.transmat_)
print(hmm2.means_)
print(hmm2.covars_)
```
hmm1 has startprob [0.99, 0.01] whereas hmm2 has startprob [0.5, 0.5].

---

As an aside, I do not know why you say that the lengths options is only available for GMMHMM?

**Comment by dizcza** [comment]
> As an aside, I do not know why you say that the lengths options is only available for GMMHMM?

I meant to say that the *chunking* via lengths is only available for GMMHMM. That is, the users could split the data if it doesn't fit into memory requirements of the algorithm, and I thought it won't make any sense for other GMM* classes because I thought that GaussianHMM and MultinomialHMM have similar issues of calling `np.vstack`. But when I looked at their implementation, now I see that in fact it's true for them as well: the statistics are being accumulated in the summation manner in the `_accumulate_sufficient_statistics` function. I can correct the comment in the docs to say it's true for all models.

I'll look into startsprob thingy.

**Comment by dizcza** [comment]
Intuitively, if the batch size is of several periods of the data stream, then everything should converge...

If you think that mentioning chunking is a bad idea, I can remove it from the docs tutorial. But the optimization I made to the M-step function is irrelevant to the batching discussion.

**Comment by dizcza** [comment]
Done.

**Comment by anntzer** [comment]
This looks good to me, but the chunking-related test should also be deferred to the "other" PR.  Also, do you want to squash the commits yourself or should I do it?
Finally, I'll ping @fcostin who was last to work on this, in case he has some time to double check the math.  (If he doesn't reply I'll merge without him, but it's always good to have an extra pair of eyes...)

**Review by fcostin — COMMENTED** [review]
I've read through the proposed patch -- it seems to be a relatively clean transformation to avoid allocating shape `(n_samples, n_components, n_mix, n_features)` arrays, and instead process per-sequence, so only `(n_samples_per_sequence, n_components, n_mix, n_features)` memory is required.

I don't understand all the math in this GHMMHMM code but the reorganisation of the algebra in this patch looks good to me.

I've made suggestions to add / remove / update comments to hopefully make what is going on easier for future readers to understand, and also suggested a couple of ways to improve the new test.

**Comment by fcostin** [comment]
For clarity, maybe this PR can now be re-titled as "reduce memory consumption during GHMMHMM multi sequence fits" -- to make it clear that the scope is now intended as a fully-backwards compatible memory optimisation.


For a PR that is focusing on optimisation, it'd be great to see some benchmarks -- perhaps of both running time and peak memory use.  Ideally we could see benchmarks on a few different scenarios. E.g. scenario where there is only a single sequence, scenario where data is split into a few large sequences, scenerio where data is split into many tiny sequences.

One thing I wondered while reviewing the code is what would happen if there are many tiny sequences.  Could this new code run a lot slower in that case, as the outer loop over sequences is now in pure-python, not executed by numpy's native code backends?  Or would that also now run much faster due to more frugal memory allocation.

**Comment by fcostin** [comment]
> Having this done (independently of the discussion above), the memory footprint will be reduced to `O(max(size_x * n_features, batch_size * n_features * n_mix * n_components))`

> But let's discuss this separately. Because to do this, I'll need to take the _do_mstep's for-loops I've introduced in this PR and move them to the _accumulate_sufficient_statistics function. These are two steps, both heavy, that's why I've decided to split the optimization in two steps for the sake of easy PR review.

@dizcza  :+1: sounds like a good improvement and great idea to split the work across multiple PRs

File: lib/hmmlearn/hmm.py [hunk 1]
```diff
@@ -1011,69 +995,76 @@ def _do_mstep(self, stats):
...
                 alphas = self.covars_prior
                 betas = self.covars_weight
+                centered_means2 = centered_means ** 2
+
+                c_n = lambdas[:, :, None] * centered_means2 + 2 * betas
+                for post_comp_mix, samples in zip(stats['post_comp_mix'],
+                                                  stats['samples']):
+                    centered = samples[:, None, None, :] - means_before_update
+                    centered2 = np.square(centered, out=centered)  # reuse
+                    c_n += np.einsum('ijk,ijkl->jkl', post_comp_mix, centered2)
```
Comment (fcostin) [inline lib/hmmlearn/hmm.py:1031]: Comment only, out of scope for this PR:

If the goal is to reduce peak memory use, it seems that there's still some waste here as we construct a huge intermediate `centered` value of shape `(n_samples_in_this_sequence, n_components, n_mix, n_features` and then do a single pass over it, immediately reducing away the `n_samples_in_this_sequence` dimensions in the `einsum`.

I reckon the calculation here is is equivalent to this naive bunch of C for loops, that doesn't need to allocate any memory for a huge `centered` array at all:

```
int i, j, k, l;
double acc, centered;

for (j = 0; j < n_components; ++j) {
    for (k = 0; k < n_mix; ++k) {
        for (l = 0; l < n_features; ++l) {
            acc = 0.0;
            for (i = 0; i < n_samples_in_this_sequence; ++i) {
                centered = samples[i, l] - means_before_update[j, k, l];
                acc += post_comp_mix[i, j, k] *  centered * centered;
            }
            c_n[j, k, l] += acc;
        }
    }
}
```

We could implement the logic by writing it in Cython, at some cost to maintainability.  I am not sure there is a clean way to express it in fast pure numpy code, as `einsum` doesn't let us use subtraction in a big tensor expression.

I'm not sure if similar idea could be used to rewrite the other cases of covariance_type.

File: lib/hmmlearn/hmm.py [hunk 2]
```diff
@@ -904,7 +904,6 @@ def _compute_log_likelihood(self, X):
 
     def _initialize_sufficient_statistics(self):
         stats = super()._initialize_sufficient_statistics()
-        stats['n_samples'] = 0
```
[the remark this answers is not in the document]
Comment (dizcza) [reply lib/hmmlearn/hmm.py:907]: Exactly.

File: lib/hmmlearn/hmm.py [hunk 3]
```diff
@@ -936,7 +935,6 @@ def _initialize_sufficient_statistics(self):
         # make memory consumption independent of the number of samples.
         stats['post_comp_mix'] = []
         stats['samples'] = []
-        stats['centered'] = []
```
[the remark this answers is not in the document]
Comment (dizcza) [reply lib/hmmlearn/hmm.py:939]: The answer is rather embarrassing: the `centering` keyword should never have been added in the stats in the first place - it's easily computed by subtracting the means from the tensor `X` (now `stats['samples']`) directly in the `_do_mstep` function.
Therefore, I leave out any comments on why I've fixed something that should never take place.

File: lib/hmmlearn/hmm.py (hunk 1 above)
Comment (dizcza) [reply lib/hmmlearn/hmm.py:1031]: I avoid any Cython optimizations - they become hard to maintain in a year, especially for outsiders.
Feel free to create a dedicated PR.

File: lib/hmmlearn/tests/test_gmm_hmm_new.py [hunk 4]
```diff
@@ -198,6 +200,34 @@ class TestGMMHMMWithFullCovars(GMMHMMTestMixin):
...
+
+        for covtype in COVARIANCE_TYPES:
+            model1 = GMMHMM(n_components=3, n_mix=2, covariance_type=covtype)
+            model2 = GMMHMM(n_components=3, n_mix=2, covariance_type=covtype)
+
+            np.random.seed(1)
+            model1.fit(data)
+
+            np.random.seed(1)
+            model2.fit(data, lengths=[100] * 10)
```
[the remark this answers is not in the document]
Comment (dizcza) [reply lib/hmmlearn/tests/test_gmm_hmm_new.py:219]: Unfortunately, fixing the `random_state` is not enough. For some reason, I still receive different answers from trial to trial.

**Comment by dizcza** [comment]
> One thing I wondered while reviewing the code is what would happen if there are many tiny sequences. Could this new code run a lot slower in that case, as the outer loop over sequences is now in pure-python, not executed by numpy's native code backends? Or would that also now run much faster due to more frugal memory allocation.

For a large number of small chunks, of course, the faster approach would be to merge all chunks in one array. I've voted for the memory footprint here, and generally, it's a memory VS speed tradeoff question. The only solution is to add the `optimization='memory'` parameter in the GMMHMM constructor ('memory' and 'speed' are valid options) to let the user decide. Alternatively, we can try-catch memory overflow errors but it doesn't seem right to me.

----

I'm done with the rest of the comments.

**Comment by anntzer** [comment]
LGTM, but you'll need to rebase to fix merge conflicts.

**Comment by dizcza** [comment]
@anntzer  The GitHub tells me that "This branch has no conflicts with the base branch"

**Comment by fcostin** [comment]
thank you @dizcza for making changes

**Comment by anntzer** [comment]
I squashed your commits and merged them as https://github.com/hmmlearn/hmmlearn/commit/44a2b681d023918fbb7d0842ff3b5a98441fbc16.  Thanks!

**Comment by dizcza** [comment]
As a matter of fact, the size of a chunk of `X` hardly influences the execution time of the `_do_mstep()` function. In other words, the for-loops I've introduced in this PR aren't that slow compared to computation done with a squashed array (as previously).

fig

To read the plot, compare the execution time at any point of the x-axis (the number of chunks in which the input array `X` is divided) with `x=1` (a single chunk), which corresponds to the full input array.

Benchmark script:

```python
from copy import deepcopy

import matplotlib.pyplot as plt
import numpy as np
import perfplot

from lib.hmmlearn.hmm import GMMHMM
from lib.hmmlearn.tests.test_gmm_hmm import create_random_gmm


timings = dict()

n_chunks = [1, 5, 10, 20, 50, 100, 200, 500, 1000]

n_samples = 10_000
n_components = 10
n_mix = 5

for covtype in ["diag", "spherical", "tied", "full"]:
    np.random.seed(0)


    gmm = create_random_gmm(n_components, n_mix, covariance_type=covtype, prng=0)
    gmm.covariances_ = gmm.covars_
    data = gmm.sample(n_samples=n_samples)[0]

    model = GMMHMM(n_components=n_components, n_mix=n_mix, covariance_type=covtype,
                   random_state=1)

    stats = {
        "start": np.random.rand(n_components),
        "trans": np.random.rand(n_components, n_components),
        "post_mix_sum": np.random.rand(n_components, n_mix),
        "post_sum": np.random.rand(n_components),
    }
    model._init(data)


    def prepare(n):
        chunk_size = n_samples // n
        stats_clone = deepcopy(stats)
        stats_clone['post_comp_mix'] = list(np.random.rand(n, chunk_size, n_components, n_mix))
        stats_clone['samples'] = list(np.random.rand(n, chunk_size, n_mix))
        return stats_clone


    bench = perfplot.bench(
        setup=prepare,  # or setup=np.random.rand
        kernels=[model._do_mstep],
        n_range=n_chunks,
        xlabel="# chunks",
    )
    timings[covtype] = bench.timings_s[0]
    print(timings[covtype])

fig, ax = plt.subplots()
for covtype in timings.keys():
    ax.plot(n_chunks, timings[covtype], label=covtype)
ax.set_xlabel('# chunks')
ax.set_ylabel('Execution time, sec')
plt.legend()
plt.tight_layout()
plt.show()
```
